packages/hardhat-core/src/internal/hardhat-network/stack-traces/vm-debug-tracer.ts

```diff
@@ -319,9 +323,9 @@ export class VMDebugTracer {
       .toString("hex")
       .match(/.{1,64}/g);
 
-    const result = memory === null ? [] : [...memory];
-
-    return result;
+    return memory === null
+      ? []
+      : memory?.slice(0, Number(step.memoryWordCount)); // Remove the additional non allocated memory
```

Comment:
if `memory` is not `null`, [it has to be an array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match#return_value). You can drop the optional chaining.